### PR TITLE
[New] Add RFC formatting support (#173)

### PR DESCRIPTION
**Pull request by Max-Kolodezniy**

### CODE REVIEW

File: lib/stringify.js [hunk 1]
```diff
@@ -24,10 +24,16 @@ var defaults = {
         return toISO.call(date);
     },
     skipNulls: false,
-    strictNullHandling: false
+    strictNullHandling: false,
+    formatters : {
+        RFC3986 : function (value) { return value; },
+        /** @return {string} */
+        RFC1738 : function (value) { return value.toString().replace(/%20/g, '+'); }
```
Comment (ljharb) [inline lib/stringify.js]: let's use `String(value)` here - also, what happens if `value` is null or undefined?

File: test/stringify.js [hunk 2]
```diff
@@ -502,4 +502,25 @@ test('stringify()', function (t) {
...
+    });
+
+    t.test('RFC 3986 spaces serialization', function (st) {
+        st.equal(qs.stringify({ a: 'b c' }, { format: 'RFC3986' }), 'a=b%20c');
+        st.end();
+    });
+
+    t.test('Backward compatibility to RFC 3986 and edge cases', function (st) {
+        st.equal(qs.stringify({ a: 'b c' }, { format: 'UFO1234' }), 'a=b%20c');
+        st.equal(qs.stringify({ a: 'b c' }, { format: false }), 'a=b%20c');
```
Comment (ljharb) [inline test/stringify.js]: I'm wondering if it would be better to throw immediately if the provided format is invalid.

Comment (ljharb) [inline test/stringify.js]: Specifically, so if someone passes a specific RFC that we don't yet support, it doesn't silently pass.

Comment (Max-Kolodezniy) [inline test/stringify.js]: Agree, will fix that today

File: lib/stringify.js (hunk 1 above)
Comment (Max-Kolodezniy) [inline lib/stringify.js]: Roger that

**Comment by Max-Kolodezniy** [comment]
Unfortunately, I cannot fix all Lint errors in this stringily without deep refactoring. I'd like to postpone it from now and change a little bit the .eslint file

**Comment by slavafomin** [comment]
Great job, thanks!

However, I would suggest to pass format not as a string, but by referencing one of the predefined constants, like in [my example](https://github.com/ljharb/qs/issues/173#issuecomment-251311153).

We could put those under a property for namespacing, e.g. `qs.formats.RFC1738`.

Such approach would allow IDE's to highlight errors and to provide code completion when using the module.

**Comment by Max-Kolodezniy** [comment]
You are right, IDE autocomplete support would be great. On my way to commit that.

File: lib/index.js [hunk 3]
```diff
@@ -5,5 +5,9 @@ var parse = require('./parse');
 
 module.exports = {
     parse: parse,
+    // String constants
```
Comment (ljharb) [inline lib/index.js]: let's group these under a `formats` object, so it's clear that "parse" isn't a format, and so the full list can be obtained via reflection.

File: lib/stringify.js [hunk 4]
```diff
@@ -20,14 +20,20 @@ var defaults = {
     delimiter: '&',
     encode: true,
     encoder: utils.encode,
+    format: 'RFC3986',
+    formatters: {
+        /** @return {string} */
```
Comment (ljharb) [inline lib/stringify.js]: please remove this; i don't want jsdoc anywhere near code that i have to maintain :-)

File: lib/stringify.js [hunk 5]
```diff
@@ -43,9 +49,9 @@ var stringify = function stringify(object, prefix, generateArrayPrefix, strictNu
 
     if (typeof obj === 'string' || typeof obj === 'number' || typeof obj === 'boolean' || utils.isBuffer(obj)) {
         if (encoder) {
-            return [encoder(prefix) + '=' + encoder(obj)];
+            return [encoder(prefix) + '=' + formatter(encoder(obj))];
```
Comment (ljharb) [inline lib/stringify.js]: does the formatter only need to apply to values, or should it also apply to keys?

File: lib/stringify.js [hunk 6]
```diff
@@ -112,6 +120,12 @@ module.exports = function (object, opts) {
     var sort = typeof options.sort === 'function' ? options.sort : null;
     var allowDots = typeof options.allowDots === 'undefined' ? false : options.allowDots;
     var serializeDate = typeof options.serializeDate === 'function' ? options.serializeDate : defaults.serializeDate;
+    if (typeof options.format === 'undefined') {
+        options.format = defaults.format;
+    } else if (!(options.format in defaults.formatters)) {
+        throw new TypeError('Unknown format options provided. Valid values are RFC3986 and RFC1738.');
```
Comment (ljharb) [inline lib/stringify.js]: Probably instead of a hardcoded list here, we should get the list off of `qs.formats`, or omit it entirely?

File: lib/stringify.js (hunk 4 above)
Comment (slavafomin) [inline lib/stringify.js]: What's wrong with jsdoc? It's very helpful with IDE code autocompletion and code proofing.

File: lib/stringify.js (hunk 6 above)
Comment (ljharb) [inline lib/stringify.js]: Sounds good!

Comment (Max-Kolodezniy) [inline lib/stringify.js]: We have to notify user if provided format will not work. So, reaction to the wrong argument provided should be straight and self-explanatory

File: lib/stringify.js (hunk 5 above)
Comment (Max-Kolodezniy) [inline lib/stringify.js]: It's not common but possible. Will add that

File: lib/stringify.js (hunk 6 above)
Comment (Max-Kolodezniy) [inline lib/stringify.js]: Also, qs.formats is unavailable in stringify.js. And including index.js will cause sort of endless inclusion cycle

File: lib/index.js [hunk 7]
```diff
@@ -4,6 +4,10 @@ var stringify = require('./stringify');
 var parse = require('./parse');
 
 module.exports = {
+    formats: {
```
Comment (ljharb) [inline lib/index.js]: could this be stored in a separate file, and required in here? that way `stringify.js` could require it, and use the constant rather than the string

File: lib/stringify.js [hunk 8]
```diff
@@ -43,9 +48,9 @@ var stringify = function stringify(object, prefix, generateArrayPrefix, strictNu
 
     if (typeof obj === 'string' || typeof obj === 'number' || typeof obj === 'boolean' || utils.isBuffer(obj)) {
         if (encoder) {
-            return [encoder(prefix) + '=' + encoder(obj)];
+            return [formatter(encoder(prefix)) + '=' + formatter(encoder(obj))];
         }
-        return [prefix + '=' + String(obj)];
+        return [formatter(prefix) + '=' + formatter(String(obj))];
```
Comment (ljharb) [inline lib/stringify.js]: i'm wondering if we should pass a "key" and "value" string argument to `formatter` as well, in case the user wants to format them differently

File: lib/stringify.js [hunk 9]
```diff
@@ -112,6 +119,12 @@ module.exports = function (object, opts) {
     var sort = typeof options.sort === 'function' ? options.sort : null;
     var allowDots = typeof options.allowDots === 'undefined' ? false : options.allowDots;
     var serializeDate = typeof options.serializeDate === 'function' ? options.serializeDate : defaults.serializeDate;
+    if (typeof options.format === 'undefined') {
+        options.format = defaults.format;
+    } else if (!(options.format in defaults.formatters)) {
```
Comment (ljharb) [inline lib/stringify.js]: this will allow a format of `toString`, for example - instead of `in`, could we use `Object.prototype.hasOwnProperty.call`?

File: test/stringify.js [hunk 10]
```diff
@@ -502,4 +502,35 @@ test('stringify()', function (t) {
...
+        st.equal(qs.stringify({ a: 'b c' }), 'a=b%20c');
+        st.end();
+    });
+
+    t.test('Edge cases and unknown formats', function (st) {
+        ['UFO1234', false, 1234, null, {}, []].forEach(
+            function (format) {
+                st.throws(
+                    function () {
+                        qs.stringify({ a: 'b c' }, {format: format});
```
Comment (ljharb) [inline test/stringify.js]: object curly braces need spaces inside them

File: lib/stringify.js (hunk 8 above)
Comment (Max-Kolodezniy) [inline lib/stringify.js]: I don't see how it can be formatted differently if user provide only one option. Formatter is dedicated function for string variables - keys and values and "=" sign and concatenation is not an formatter responsibility.

File: lib/index.js (hunk 7 above)
Comment (Max-Kolodezniy) [inline lib/index.js]: Hm, include the file for two strings gives uncleared and unjustified overhead. And in the same time - it replaces the only one occurrence in stringify.js : input parameter validation.
And mix constants and formatters to the one file will be out of KISS

File: test/stringify.js (hunk 10 above)
Comment (Max-Kolodezniy) [inline test/stringify.js]: Fixed

File: lib/index.js (hunk 7 above)
Comment (ljharb) [inline lib/index.js]: What overhead? Modularity is far more valuable than avoiding a function call.

**Comment by ljharb** [comment]
I think with the final change of moving the formats to a separate file, plus a rebase on top of latest master (ideally squashed down to a commit), this should be good to go.

**Comment by Max-Kolodezniy** [comment]
I'm out of PC this week, travelling with smartphone only. Will do that in a
week

10 жовт. 2016 р. 9:12 пп "Jordan Harband" <notifications@github.com> пише:

File: lib/stringify.js (hunk 4 above)
Comment (slavafomin) [inline lib/stringify.js]: How could IDE parse the types of the variables, for example, when JavaScript doesn't have strict types?

**Comment by Max-Kolodezniy** [comment]
This code has been changed. Now it's encapsulated in the separate file

12 жовт. 2016 р. 1:08 пп "Slava Fomin II" <notifications@github.com> пише:

**Comment by ljharb** [comment]
@Max-Kolodezniy thanks! I think it's ready to go once it gets rebased down to a single commit. Since you're traveling, I'm happy to take care of it for you if you don't mind?

**Comment by Max-Kolodezniy** [comment]
Feel free to do that!

12 жовт. 2016 р. 11:13 пп "Jordan Harband" <notifications@github.com> пише:
